### PR TITLE
Telemetry and bug fix extravaganza

### PR DESCRIPTION
**Pull request by SmilyOrg**
1. Telemetry now covers more things, notably the (SDL) event loop.
2. Added support for unsigned 64bit integers to NetworkBuffer.
3. Better buffer usage in AssetProtocolHandler.
4. AssetProtocolHandlers can now be freed properly.
5. loom_newArray allocation now skips constructors for fundamental types.
6. Telemetry now uses an event-and-lookup design, which is a bit tougher on bandwidth, but much lower on processing. The bandwidth issue is side-stepped with new buffered network writes.
7. Network writes are now buffered in a bip buffer. This avoids stalls and blocking.
8. The asset agent connection is now more aware of issues on both ends, reads and writes now time out after some time and the connection is dropped.
9. Changed from double to long long in readTimerNano to avoid unnecessary FP ops.
10. Fixed some JSON leaks and simplified implementation / memory management wrt usage a bit.
11. Added a `useCache` parameter to `utByteArray` to expose the underlying `utArray` cache.
12. Added `utByteArray` reading/writing unsigned 64-bit ints.
13. Added temp fix for reading/writing unaligned `utByteArray` strings.
14. Added uthash, a simple hash table for use from C code.
15. Added `utString::alloc` to provide a way to reserve a string buffer in advance.
16. Added an unsigned 64-bit int hash key support to `utHashTable`.
17. Changed some SQLite code to use the newer JSON APIs that manage memory automatically.
18. Fixed texture mipmapping code to actually allocate properly.
19. Fixed some internal string functions to use the lua string length and not recompute all the time with strlen, noticable with larger strings.
20. Fixed a bool warning for `openURL`.
21. Made a `MemberInfo` check function faster by avoiding string length computation.
22. Added `LoomProfilerRoot` to `MethodBase` to support the new telemetry system.
23. Fixed some telemetry server memory leaks.
24. Added a ping timeout to the asset agent connection.
25. Added support for multiple "root" tick ranges to the telemetry client.
26. Added display of write/read overhead to telemetry client.

### CODE REVIEW

File: application/android/src/co/theengine/loomplayer/LoomHTTP.java [hunk 1]
```diff
@@ -185,17 +185,14 @@ public void onSuccess(int statusCode, Header[] headers, byte[] binaryData)
...
                         BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(savedFile));
                         bos.write(binaryData);
                         bos.flush();
                         bos.close();
-                        Log.d(TAG, "File written...");
                     }
                     catch (Exception e)
                     {
-                        Log.e(TAG, "File write failed...");
                         throw new AssertionError("HTTP Response could not be cached.");
```
Comment (bengarney) [inline application/android/src/co/theengine/loomplayer/LoomHTTP.java]: Should we still log this somehow?

File: loom/common/core/performance.cpp [hunk 2]
```diff
@@ -233,14 +233,6 @@ void LoomProfiler::reset()
 
 LoomProfilerRoot::LoomProfilerRoot(const char *name)
 {
-    for (LoomProfilerRoot *walk = sRootList; walk; walk = walk->mNextRoot)
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: Would this be good to keep in debug build?

File: loom/common/core/performance.cpp [hunk 3]
```diff
@@ -252,8 +244,20 @@ LoomProfilerRoot::LoomProfilerRoot(const char *name)
     mTotalInvokeCount       = 0;
     mFirstLoomProfilerEntry = NULL;
     mEnabled                = true;
+    mTelemetryVisited       = false;
 }
 
+LoomProfilerRoot* LoomProfilerRoot::fromName(const char *name)
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: Is this used for anything important?

File: loom/common/core/telemetry.cpp [hunk 4]
```diff
@@ -17,17 +18,22 @@ int Telemetry::tickId = 0;
...
+int Telemetry::tickThreadId = -1;
 
 // Initialize specialized constants for every type used
 // TickMetricValue
 template<> const TableType TableValues<TickMetricValue>::type = 1;
 template<> const int TableValues<TickMetricValue>::packedItemSize = 4 + 8;
 // TickMetricRange
 template<> const TableType TableValues<TickMetricRange>::type = 2;
-template<> const int TableValues<TickMetricRange>::packedItemSize = 4 + 4 * 4 + 2 * 8;
+template<> const int TableValues<TickMetricRange>::packedItemSize = 4 + 4 * 4 + 3 * 8;
```
Comment (bengarney) [inline loom/common/core/telemetry.cpp]: Why this change? Can we generate this from constants or note what the magic #s are?

File: loom/common/core/telemetry.cpp [hunk 5]
```diff
@@ -42,133 +48,277 @@ void Telemetry::disable()
...
 
-    lmAssert(tickTimerStack.size() == 0, "Tick timer end call missing");
+    size_t eventsSize = buffer.readUnsignedInt();
+    unsigned int eventEndianness = buffer.readUnsignedInt();
+    size_t eventsPos = buffer.getPosition();
+    
+    // Skip to the profiler root tree
+    buffer.setPosition(eventsPos + eventsSize);
+
+    utHashTable<utUInt64HashKey, TickProfilerRoot> addrToName;
```
Comment (bengarney) [inline loom/common/core/telemetry.cpp]: Can we reserve this and/or ancestors to avoid unnecessary growing of the structures?

File: application/android/src/co/theengine/loomplayer/LoomHTTP.java (hunk 1 above)
Comment (SmilyOrg) [inline application/android/src/co/theengine/loomplayer/LoomHTTP.java]: I think the error trickling up the stack should be enough?

File: loom/common/platform/platformNetwork.c [hunk 6]
```diff
@@ -61,13 +66,68 @@ typedef int   SOCKET;
 
 lmDefineLogGroup(netLogGroup, "net", 1, LoomLogWarn);
 
+static const int readSpinSleepMs = 5;
```
Comment (bengarney) [inline loom/common/platform/platformNetwork.c]: Should we comment what these are? They seem important.

File: loom/common/platform/platformNetwork.c [hunk 7]
```diff
@@ -61,13 +66,68 @@ typedef int   SOCKET;
...
+
+    // The number of bytes sent, used for debug log reporting
+    unsigned int bytesSent;
+
+    // Required structure for the hash to work
+    UT_hash_handle hh;
+};
+
+// Head of the socket id -> SocketData hash table
+static struct SocketData* socketDatas = NULL;
```
Comment (bengarney) [inline loom/common/platform/platformNetwork.c]: Maybe name it gSocketTable or something? "Datas" is weird.

File: loom/common/platform/platformNetwork.c [hunk 8]
```diff
@@ -399,80 +476,256 @@ void loom_net_readTCPSocket(loom_socketId_t s, void *buffer, int *bytesToRead, i
...
-            break;
+            sd->stallWrites = 0;
+            sd->stalled = 0;
         }
-#endif
 
-        if (loom_net_isSocketDead(s))
+        // Debug print log output
+        if (print) {
+            platform_debugOut("Socket %x%s write buffer status: %d %s/s, %lld%% full, %d left, %d free, %d total",
```
Comment (bengarney) [inline loom/common/platform/platformNetwork.c]: Should this be lmLog instead of debugout?

File: loom/common/core/performance.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: Maybe. Due to how the new system reuses roots through this exact same search below, we shouldn't get duplicates unless we add some additional code that creates new roots manually.

File: tools/assetAgent/src/main.cpp [hunk 9]
```diff
@@ -754,10 +756,22 @@ static int socketListeningThread(void *payload)
...
+			
             for (UTsize i = 0; i < gActiveHandlers.size(); i++)
             {
-                gActiveHandlers[i]->process();
+				AssetProtocolHandler* aph = gActiveHandlers[i];
+                aph->process();
+
+                // Check for ping timeout
+				int msSincePing = loom_readTimer(aph->lastActiveTime);
+				if (msSincePing > socketPingTimeoutMs)
```
Comment (bengarney) [inline tools/assetAgent/src/main.cpp]: Bad indent in this file?

File: loom/common/core/performance.cpp (hunk 3 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: It's how we retrieve the correct profiler root from a function name now. It's possible that it's slow, however not any slower than before as the same check happened in the constructor above.

File: application/android/src/co/theengine/loomplayer/LoomHTTP.java (hunk 1 above)
Comment (bengarney) [inline application/android/src/co/theengine/loomplayer/LoomHTTP.java]: Ah, it does rethrow. Fine with me.

File: loom/common/core/performance.cpp (hunk 2 above)
Comment (bengarney) [inline loom/common/core/performance.cpp]: But this was there to catch dupes on the C++ macros. Ie PROFILE_START(a); PROFILE_START(a); Do we still catch that elsewhere?

File: loom/common/core/performance.cpp (hunk 3 above)
Comment (bengarney) [inline loom/common/core/performance.cpp]: Oh OK. Do we do it just first time we encounter that MethodInfo or every time?

File: loom/common/core/telemetry.cpp (hunk 4 above)
Comment (SmilyOrg) [inline loom/common/core/telemetry.cpp]: I think this part of the code isn't even used anymore and I should probably remove everything related to it before landing.

File: loom/common/core/telemetry.cpp (hunk 5 above)
Comment (SmilyOrg) [inline loom/common/core/telemetry.cpp]: Reserving this would be a good idea, yes. Maybe we should add a size at the beginning so we can preallocate completely. We can reserve ancestors, but I don't think it'll have much of an effect as it's only as big as the tree depth.

Both of these are eclipsed by JSON allocation / processing / generation though.

File: loom/common/core/performance.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: Hmm, you're right, this will have to be handled. Maybe we can take advantage of `stringtable_insert` or, a probably better idea, just use a `utHashTable` to keep track of all the roots. Might make things faster as well.

File: loom/common/core/performance.cpp (hunk 3 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: First time:
```
    if (method->profilerRoot == NULL)
    {
        method->profilerRoot = LoomProfilerRoot::fromName(method->getFullMemberName());
    }
```

File: loom/common/platform/platformNetwork.c (hunk 6 above)
Comment (SmilyOrg) [inline loom/common/platform/platformNetwork.c]: Yes, I forgot to.

File: loom/common/platform/platformNetwork.c (hunk 7 above)
Comment (SmilyOrg) [inline loom/common/platform/platformNetwork.c]: Yeah I didn't know what to call it and it got stuck. Socket table is good, maybe socket hash since it's really sort of a funky hash table.

File: loom/common/platform/platformNetwork.c (hunk 8 above)
Comment (SmilyOrg) [inline loom/common/platform/platformNetwork.c]: Not without additional work on the logging system. All the log messages get sent through a socket, so if you try to send logging logs of logging while logging you'll have a bad time.

File: tools/assetAgent/src/main.cpp (hunk 9 above)
Comment (SmilyOrg) [inline tools/assetAgent/src/main.cpp]: Derp, I had to reset my VS settings, so this looks like a casualty of that.

File: loom/common/core/performance.cpp (hunk 3 above)
Comment (bengarney) [inline loom/common/core/performance.cpp]: OK Cool.

**Comment by SmilyOrg** [comment]
Ok, this is ready for another look!

File: loom/common/core/performance.cpp [hunk 10]
```diff
@@ -233,13 +239,10 @@ void LoomProfiler::reset()
...
 {
-    for (LoomProfilerRoot *walk = sRootList; walk; walk = walk->mNextRoot)
-    {
-        if (!strcmp(walk->mName, name))
-        {
-            lmAssert(false, "Duplicate profile name: %s", name);
-        }
-    }
+    LookupType* lookup = static_cast<LookupType*>(sRootLookup);
+    utHashedString key(name);
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: Is this going to cause problems with collisions in hash?

File: loom/common/core/performance.cpp [hunk 11]
```diff
@@ -273,7 +293,7 @@ void LoomProfiler::validate()
                 }
             }
 
-            lmAssert(wk, "Validation failed - couldnot find child in its parent's list.");
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: Semi-contraction. Wouldnot mightnot havenot.

File: loom/common/platform/platform.cpp [hunk 12]
```diff
@@ -28,6 +28,8 @@ extern "C" {
 
 int platform_openURL(const char *url)
 {
+// Casting to int here is the right thing to do based on the func docs
+#pragma warning ( suppress: 4311, 4302 )
```
Comment (bengarney) [inline loom/common/platform/platform.cpp]: Can we pop warning suppress after?

File: loom/common/platform/platformNetwork.c (hunk 7 above)
Comment (bengarney) [inline loom/common/platform/platformNetwork.c]: socketTable or socketHashTable are good.

File: loom/common/platform/platformNetwork.c (hunk 8 above)
Comment (bengarney) [inline loom/common/platform/platformNetwork.c]: Can we comment that? It's a subtle but important point.

File: loom/common/core/performance.cpp (hunk 10 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: Not after merging with that string hash fix that also checks for collisions by comparing the exact strings.

File: loom/common/core/performance.cpp (hunk 11 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: I thought that was actually a thing -_-

File: loom/common/platform/platform.cpp (hunk 12 above)
Comment (SmilyOrg) [inline loom/common/platform/platform.cpp]: Suppress is push, disable and pop in one statement!

Source: http://stackoverflow.com/a/18461418

**Comment by SmilyOrg** [comment]
How does this stand now?

File: loom/common/core/performance.cpp (hunk 11 above)
Comment (bengarney) [inline loom/common/core/performance.cpp]: Only with the apostrophe! Logically it would work, though.

**Comment by bengarney** [comment]
I find this.... suitable.
